torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,305 @@
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+        Node* n = *it;
+        it++; // advance iterator bc the current node may be destroyed
+
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::GetAttr) {
+          auto name = n->s(attr::name);
+          auto attrModule = module_;
+          auto inputNode = n->inputs()[0]->node();
+          if (!findConstantAttr(inputNode, name, attrModule)) {
+            GRAPH_DEBUG("attribute: ", name, " is mutable.")
+            continue;
+          }
+          assert(attrModule.hasattr(name));
+          Value* paramConst = nullptr;
+          auto I = attrValues.find(attrModule._ivalue());
+          if (I != attrValues.end()) {
+            auto II = I->second.find(name);
+            if (II != I->second.end())
+              paramConst = II->second;
+          }
+          if (!paramConst) {
+            auto attr = attrModule.attr(name);
+            if (!shouldFoldAttr(attr, name, isEval))
+              continue;
+            if (auto attrVal = tryInsertConstant(*graph, attr)) {
+              paramConst = *attrVal;
+            } else {
+              GRAPH_DEBUG(
+                  attr.type()->cast<ClassType>() ? "" : "attribute: ",
+                  name,
+                  " is not materializable.");
+              continue;
+            }
+            auto moduleName = attrModule.type()->name()->qualifiedName();
+            moduleName += ".";
+            moduleName += name;
+            paramConst->setDebugName(moduleName);
+            attrValues[attrModule._ivalue()][name] = paramConst;
+          }
+          GRAPH_UPDATE(
+              "Folding GetAttr %",
+              n->outputs()[0]->debugName(),
+              " with ",
+              paramConst->debugName());
+          n->outputs().at(0)->replaceAllUsesWith(paramConst);
+          n->removeAllInputs();
+        }
+      }
+    }
+  }
+
+  // cleanupFrozenModule function cleans up the Frozen module. it performs the
+  // following:
+  // 1) Remove unused attributes.
+  // 2) Remove unreferenced submodules
+  // 3) Remove non pulic unreferenced methods.
+  // TODO: do #3 because there is no API to 'unsafely' remove methods.
+  void cleanupFrozenModule(std::shared_ptr<Graph>& graph) {
+    std::vector<std::string> attrsToRemove;
+    auto type = module_.type();
+    size_t N = type->numAttributes();
+    auto KeepAttrs = recordReferencedAttrs(graph);
+    for (size_t i = 0; i < N; ++i) {
+      auto attrTy = type->getAttribute(i);
+      auto name = type->getAttributeName(i);
+      if (!KeepAttrs.count(name)) {
+        attrsToRemove.push_back(name);
+      }
+    }
+    for (auto& name : attrsToRemove) {
+      module_._ivalue()->unsafeRemoveAttr(name);
+      module_.type()->unsafeRemoveAttribute(name);
+    }
```

Comment:
I think you probably need to extend this for shared ClassType, similar to https://github.com/pytorch/pytorch/pull/31408

Reply:
I changed one of my testcases to verify current implementation works fine. The current patch cleans up only top level module. submodules that contains mutable attributes are  not clean up. 

I plan  to implement a follow up patch to do that and that case, shared ClassType need to be carefully handled.